x/distribution/spec/04_messages.md

```diff
@@ -27,163 +30,91 @@ func (k Keeper) SetWithdrawAddr(ctx sdk.Context, delegatorAddr sdk.AccAddress, w
 ```
-
-## Common calculations 
-
-### Update total validator accum
-
-The total amount of validator accum must be calculated in order to determine
-the amount of pool tokens which a validator is entitled to at a particular
-block. The accum is always additive to the existing accum. This term is to be
-updated each time rewards are withdrawn from the system. 
-
-```go
-func (g FeePool) UpdateTotalValAccum(height int64, totalBondedTokens Dec) FeePool
-    blocks = height - g.TotalValAccumUpdateHeight
-    g.TotalValAccum += totalDelShares * blocks
-    g.TotalValAccumUpdateHeight = height
-    return g
+stake := initial stake
+rewards := 0
+previous := A
+for P in P1, ..., PN`:
+    rewards = (R(P) - previous) * stake
+    stake = stake * F(P)
+    previous = P
+rewards = rewards + (R(B) - R(PN)) * stake
 ```
 
-### Update validator's accums
-
-The total amount of delegator accum must be updated in order to determine the
-amount of pool tokens which each delegator is entitled to, relative to the
-other delegators for that validator. The accum is always additive to
-the existing accum. This term is to be updated each time a
-withdrawal is made from a validator. 
+Put another way, the historical rewards are calculated retroactively by playing back all the slashes, and attenuating the delegator's stake at each step.
+The final calculated stake should be equivalent to the actual staked coins in the delegation, within a margin of error due to rounding errors.
 
-``` go
-func (vi ValidatorDistInfo) UpdateTotalDelAccum(height int64, totalDelShares Dec) ValidatorDistInfo
-    blocks = height - vi.TotalDelAccumUpdateHeight
-    vi.TotalDelAccum += totalDelShares * blocks
-    vi.TotalDelAccumUpdateHeight = height
-    return vi
-```
+Response:
 
-### FeePool pool to validator pool
++++ https://github.com/cosmos/cosmos-sdk/blob/v0.40.0/proto/cosmos/distribution/v1beta1/tx.proto#L42-L50
 
-Every time a validator or delegator executes a withdrawal or the validator is
-the proposer and receives new tokens, the relevant validator must move tokens
-from the passive global pool to their own pool. It is at this point that the
-commission is withdrawn
+## WithdrawValidatorCommission
 
-```go
-func (vi ValidatorDistInfo) TakeFeePoolRewards(g FeePool, height int64, totalBonded, vdTokens, commissionRate Dec) (
-                                vi ValidatorDistInfo, g FeePool)
-
-    g.UpdateTotalValAccum(height, totalBondedShares)
-    
-    // update the validators pool
-    blocks = height - vi.FeePoolWithdrawalHeight
-    vi.FeePoolWithdrawalHeight = height
-    accum = blocks * vdTokens
-    withdrawalTokens := g.Pool * accum / g.TotalValAccum 
-    commission := withdrawalTokens * commissionRate
-    
-    g.TotalValAccum -= accumm
-    vi.PoolCommission += commission
-    vi.PoolCommissionFree += withdrawalTokens - commission
-    g.Pool -= withdrawalTokens
-
-    return vi, g
-```
+The validator can send this message to withdraw their accumulated commission.
```

Comment:
please use something specific here instead of `this`.

Reply:
yes! add the message name and even a link to the message in the docs